mirgecom/filter.py

```diff
@@ -0,0 +1,194 @@
+.. automethod: filter_modally
+"""
+
+__copyright__ = """
+Copyright (C) 2020 University of Illinois Board of Trustees
+"""
+
+__license__ = """
+Permission is hereby granted, free of charge, to any person obtaining a copy
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in
+all copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
+"""
+
+import numpy as np
+import loopy as lp
+from grudge import sym
+from meshmode.dof_array import DOFArray
+from pytools import memoize_in
+from pytools.obj_array import (
+    obj_array_vectorized_n_args,
+    #    obj_array_vectorize
+)
+
+
+def exponential_mode_response_function(mode, alpha, cutoff, nfilt, filter_order):
+    """Return an exponential filter coefficient for the user-provided *mode*."""
+    return np.exp(-1.0 * alpha * ((mode - cutoff) / nfilt)
+                  ** (2*filter_order))
+
+
+def make_spectral_filter(element_group, cutoff, mode_response_function):
+    r"""
+    Create a spectral filter with the provided *mode_response_function*.
+
+    This routine returns a filter operator in the modal basis designed to
+    apply the user-provided *mode_response_function* to the spectral modes
+    beyond the user-provided *cutoff*.
+
+    Parameters
+    ----------
+    element_group: :class:`meshmode.mesh.MeshElementGroup`
+        A :class:`meshmode.mesh.MeshElementGroup` from which the mode ids,
+        element order, and dimension may be retrieved.
+    cutoff: integer
+        Mode cutoff beyond which the filter will be applied, and below which
+        the filter will preserve.
+    mode_response_function:
+        A function that returns a filter weight for for each mode id.
+
+    Returns
+    -------
+    filter: np.ndarray
+        filter operator in the modal basis
+    """
+    mode_ids = element_group.mode_ids()
+    order = element_group.order
+    dim = element_group.dim
+
+    nmodes = len(mode_ids)
+    filter = np.identity(nmodes)
+    nfilt = order - cutoff
+    if nfilt <= 0:
+        return filter
+
+    for mode_index, mode_id in enumerate(mode_ids):
+        mode = mode_id
+        if dim > 1:
+            mode = sum(mode_id)
```

Comment:
Why not pass on the mode ID unchanged?

Reply:
I had an error when doing 1D.  I think when calling `sum()` on an integer.

Reply:
Reminding myself now... the way I had it before, I would get an error when trying to run the filtering on 1D discretizations.  The issue was that `sum()` was called on a non-iterable. I guess that `group.mode_ids()` returns an `integer` for 1D, and an iterable  object only on numdim > 1.